File: backend/app/models/analysis.py
"""
Pydantic Analiz İstek ve Yanıt Şablonları (Models) - Pydantic v2 Uyumlu
"""
from pydantic import BaseModel
from typing import List, Dict, Any, Optional

class PlyInput(BaseModel):
    material_id: str = "T300_5208"
    angle: float = 0.0
    thickness: float = 0.125

class HoleInput(BaseModel):
    x: float = 50.0
    y: float = 25.0
    diameter: float = 6.35
    load_magnitude: float = 5000.0
    load_angle: float = 0.0
    torque: float = 0.0  # Cıvata Sıkma Torku (Nm)

class LoadCase(BaseModel):
    name: str = "Case 1"
    load_magnitude: float = 5000.0
    load_angle: float = 0.0

class AnalysisRequest(BaseModel):
    width: float = 200.0
    height: float = 100.0
    plies: List[PlyInput] = []
    holes: List[HoleInput] = []
    constraint_type: str = "fixed"
    mesh_size_global: float = 5.0
    mesh_size_hole: float = 1.0
    enable_pdm: bool = False
    failure_criterion: str = "Hashin"  # Hashin, Tsai-Wu, Puck
    load_cases: Optional[List[LoadCase]] = None

class PlyResultResponse(BaseModel):
    ply_id: int
    angle: float
    hashin_max_fi: float
    dominant_mode: str
    tsai_wu_fi: float
    mos_hashin: float
    is_failed: bool

class AnalysisResponse(BaseModel):
    layup_notation: str
    total_thickness: float
    min_mos: float
    overall_status: str
    governing_criterion: str
    critical_ply: int
    critical_angle: float
    critical_mode: str
    ply_results: List[PlyResultResponse]
    A_matrix: List[List[float]]
    B_matrix: List[List[float]]
    D_matrix: List[List[float]]
    B_nonzero: bool
    applied_load: float
    computation_time_ms: float
    mesh_summary: Dict[str, Any]
    nodes: List[List[float]]
    elements: List[List[int]]
    nodal_stresses: List[List[float]]

File: frontend/src/types/analysis.ts
export interface PlyInput {
  material_id: string;
  angle: number;
  thickness: number;
}

export interface HoleInput {
  x: number;
  y: number;
  diameter: number;
  load_magnitude: number;
  load_angle: number;
  torque?: number; // Cıvata Sıkma Torku (Nm)
}

export interface LoadCase {
  name: string;
  magnitude: number;
  angle: number;
}

export interface AnalysisRequest {
  width: number;
  height: number;
  plies: PlyInput[];
  holes: HoleInput[];
  constraint_type: string;
  mesh_size_global: number;
  mesh_size_hole: number;
  enable_pdm?: boolean; // Progressive Damage Modeling
  failure_criterion?: 'Hashin' | 'Tsai-Wu' | 'Puck';
  load_cases?: LoadCase[];
}

export interface PlyResult {
  ply_id: number;
  angle: number;
  hashin_max_fi: number;
  dominant_mode: string;
  tsai_wu_fi: number;
  mos_hashin: number;
  is_failed: boolean;
}

export interface PDMStep {
  step: number;
  load_factor: number;
  load_N: number;
  failed_matrix_elements: number;
  failed_fiber_elements: number;
  status: string;
}

export interface ProgressiveDamageResult {
  ultimate_load_N: number;
  first_ply_failure_load_N: number;
  ultimate_bearing_stress_MPa: number;
  history: PDMStep[];
  is_ultimate_failed: boolean;
}

export interface AnalysisResponse {
  layup_notation: string;
  total_thickness: number;
  min_mos: number;
  overall_status: 'PASS' | 'FAIL';
  governing_criterion: string;
  critical_ply: number;
  critical_angle: number;
  critical_mode: string;
  ply_results: PlyResult[];
  A_matrix: number[][];
  B_matrix: number[][];
  D_matrix: number[][];
  B_nonzero: boolean;
  applied_load: number;
  computation_time_ms: number;
  mesh_summary: {
    n_nodes: number;
    n_elements: number;
    n_dof: number;
    element_type: string;
  };
  nodes: number[][];
  elements: number[][];
  nodal_stresses: number[][];
  nodal_hashin_fi?: number[];
  pdm_results?: ProgressiveDamageResult;
  stress_frames?: number[][][]; // Multi-step animation stress states
  envelope_results?: any;
}

export interface MaterialOption {
  id: string;
  name: string;
  category: string;
  source: string;
  ply_thickness: number;
}

export interface CustomMaterialPayload {
  id: string;
  name: string;
  category?: string;
  source?: string;
  ply_thickness: number;
  elastic: {
    E1: number;
    E2: number;
    G12: number;
    nu12: number;
  };
  strength: {
    Xt: number;
    Xc: number;
    Yt: number;
    Yc: number;
    S12: number;
    S23?: number;
  };
}
